bee-plugin/src/handler.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    error::PluginError,
+    event::EventId,
+    grpc::{plugin_client::PluginClient, DummyEvent, ShutdownRequest, SillyEvent},
+    handshake::HandshakeInfo,
+    streamer::PluginStreamer,
+    PluginId, UniqueId,
+};
+
+use bee_event_bus::EventBus;
+
+use tokio::{
+    io::{AsyncBufReadExt, BufReader},
+    process::{Child, Command},
+    select, spawn,
+    sync::{mpsc::unbounded_channel, oneshot::Sender},
+    task::JoinHandle,
+    time::sleep,
+};
+use tonic::{transport::Channel, Request};
+
+use std::{
+    any::type_name,
+    collections::{hash_map::Entry, HashMap},
+    process::Stdio,
+    time::Duration,
+};
+
+macro_rules! spawn_streamers {
+    ($self:ident, $event_id:ident, $bus:ident, $shutdown:ident, $($event_var:pat => $event_ty:ty),*) => {{
+        match $event_id {
+            $(
+                $event_var => {
+                    let (tx, rx) = unbounded_channel::<$event_ty>();
+
+                    let client = $self.client.clone();
+
+                    spawn(async move {
+                        PluginStreamer::new(rx, $shutdown, client).run().await;
+                    });
+
+                    log::info!("registering `{}` callback for the `{}` plugin", type_name::<$event_ty>(), $self.name);
+                    $bus.add_listener_with_id(move |event: &$event_ty| {
+                        if let Err(err) = tx.send(event.clone()) {
+                            log::warn!("failed to send event: {}", err);
+                        }
+                    }, UniqueId::Plugin($self.plugin_id));
+                }
+            )*
+        }
+    }};
+}
+
+/// A handler for a plugin.
+pub(crate) struct PluginHandler {
+    /// The identifier of the plugin.
+    plugin_id: PluginId,
+    /// The name of the plugin.
+    name: String,
+    /// Shutdown for every `PluginStreamer` used by the plugin.
+    shutdowns: HashMap<EventId, Sender<()>>,
+    /// The OS process running the plugin.
+    process: Child,
+    /// The gRPC client.
+    client: PluginClient<Channel>,
+    /// The task handling stdio redirection.
+    stdio_task: JoinHandle<Result<(), std::io::Error>>,
+}
+
+impl PluginHandler {
+    /// Creates a new plugin handler from a process running the plugin logic.
+    pub(crate) async fn new(
+        plugin_id: PluginId,
+        mut command: Command,
+        bus: &EventBus<'static, UniqueId>,
+    ) -> Result<Self, PluginError> {
+        command.kill_on_drop(true).stdout(Stdio::piped()).stderr(Stdio::piped());
+
+        log::info!(
+            "spawning command `{:?}` for the plugin with ID `{:?}`",
```

Comment:
I suggest printing also the plugin name, because the ID is not fixed for a particular plugin (if I add a new plugin to the directory ids potentially change which could make debugging harder)

Reply:
here we cannot do it because the name is obtained after the handshake.